render/egl.c

```diff
@@ -347,13 +369,113 @@ struct wlr_egl *wlr_egl_create(EGLenum platform, void *remote_display,
 		}
 	}
 
+	return true;
+}
+
+struct wlr_egl *wlr_egl_create(EGLenum platform, void *remote_display,
+		const EGLint *config_attribs) {
+	struct wlr_egl *egl = egl_create();
+	if (egl == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create EGL context");
+		goto error;
+	}
+
+	if (!egl_init(egl, platform, remote_display, config_attribs)) {
+		wlr_log(WLR_ERROR, "Failed to initialize EGL context");
+		goto error;
+	}
+
 	return egl;
 
 error:
 	eglMakeCurrent(EGL_NO_DISPLAY, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
 	if (egl->display) {
 		eglTerminate(egl->display);
 	}
+	if (egl) {
+		free(egl);
+	}
+	eglReleaseThread();
+	return NULL;
+}
+
+static bool device_has_name(const drmDevice *device, const char *name) {
+	for (size_t i = 0; i < DRM_NODE_MAX; i++) {
+		if (!(device->available_nodes & (1 << i))) {
+			continue;
+		}
+		if (strcmp(device->nodes[i], name) == 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+struct wlr_egl *wlr_egl_create_from_drm_fd(int drm_fd) {
+	struct wlr_egl *egl = egl_create();
+	if (egl == NULL) {
+		wlr_log(WLR_ERROR, "Failed to create EGL context");
+		goto error;
+	}
+
+	if (egl->procs.eglQueryDevicesEXT == NULL) {
+		wlr_log(WLR_ERROR, "EGL_EXT_device_enumeration not supported");
+		goto error;
+	}
+
+	static const EGLint MAX_DEVICES = 64;
+	EGLint nb_devices = 0;
+	if (!egl->procs.eglQueryDevicesEXT(MAX_DEVICES, NULL, &nb_devices)) {
```

Comment:
The spec says:

> If devices is NULL, then max_devices will be ignored, no devices will be returned in devices, and num_devices will be set to the number of supported devices in the system.

So I think we can just drop `MAX_DEVICES` here, and call `eglQueryDevicesEXT(0, NULL, &nb_devices)`.